### PR TITLE
Feature/use webhooks

### PR DESCRIPTION
**Pull request by kebetsi**

### CODE REVIEW

File: code/src/errors/factory.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+};
+
+factory.internalError = (message: string, detail: ?string, errorForInternalUsage: ?Error): PYError => {
+  const e: Error = new Error('Internal Error');
+  const stack = e.stack.replace(/^[^\(]+?[\n$]/gm, '')
+    .replace(/^\s+at\s+/gm, '')
+    .replace(/^Object.<anonymous>\s*\(/gm, '{anonymous}()@')
+    .split('\n');
+  if (!SHUSHLOG) {
+    // console.log(errorForInternalUsage, message, detail, stack);
```
Comment (loic-pryv) [inline code/src/errors/factory.js]: Should you uncomment the console.log ?

File: code/src/fields/stream.js [hunk 2]
```diff
@@ -1,51 +1,55 @@
...
+  cache.getStreams(pryvCredentials, function (error: ?Error, streamsArray: ?Array<Stream>): void {
+    if (error != null) {
+      return next(errors.internalError('Failed fetching streams', '', error));
     }
     addStreams(0, streamsArray);
 
-    if (config.get('debug:fieldsStreams')) {
-      console.log(result);
+    if (config.get('debug:fieldsStreams') != null) {
+      //console.log(result);
```
Comment (loic-pryv) [inline code/src/fields/stream.js]: Should you uncomment the console.log ?

File: code/src/middleware/bearerAuth.js [hunk 3]
```diff
@@ -1,56 +1,66 @@
...
+  db.getSet(oauthToken, function (error: ?Error, credentials: ?Credentials) {
+    if (error != null) {
+      return next(errors.internalError('Database error', '', error));
     }
-    return next();
-  } else {
-    var authorizationHeader = req.get('Authorization').split(' ');
-    if (authorizationHeader.length !== 2) {
-      return next(PYError.authentificationRequired('Authorization header bad number of arguments'));
+    if (credentials == null ||
```
Comment (loic-pryv) [inline code/src/middleware/bearerAuth.js]: why not the "usual" way ?
_(!credentials || (!credentials.username && !credentials.urlEndpoint))_
Just asking, no need to edit...

File: code/src/routes/ifttt-v1/actions/start-stop-activity.js [hunk 4]
```diff
@@ -1,56 +1,55 @@
...
+import type { Credentials } from '../../../types';
 
 // Draft
 // - the lib-js to support / start / stop
 // - stream list should display singleActivities indicators
 
-
-module.exports = function setup(app) {
-  var actionPath = versionPath + 'actions/start-stop-activity';
+// TODO include in next release
```
Comment (loic-pryv) [inline code/src/routes/ifttt-v1/actions/start-stop-activity.js]: Let the TODO ?

File: code/src/routes/ifttt-v1/triggers/new-file.js [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
-    if (! event.attachments || event.attachments.length === 0) {
+    if (event.attachments == null || event.attachments.length === 0) {
       console.log('Skipping event ' + event.id + ' missing attachment data');
-      return false;
+      return;
     }
-    map.FileURL = event.attachmentUrl(event.attachments[0]);
+
+    const pryvCredentials = event.pryvCredentials;
+    const attachment = event.attachments[0];
```
Comment (loic-pryv) [inline code/src/routes/ifttt-v1/triggers/new-file.js]: Is it possible `event` doesn't contain `attachments` ?
If so we should check `event.attachments` existence before accessing it.

**Comment by loic-pryv** [comment]
7 tests doesn't pass. Here is the output : 

```
yarn run v1.17.3
$ env debug=false node_modules/.bin/mocha 'dist/test/**/*.test.js'


info: ready: [IFTTT Bridge v0.1.1 in development mode listening on:0.0.0.0:8080]
  /actions/new-file/
    /
info: database version :0.0.1
info: ready: [database] > Redis
[COMMENT FROM LOIC] post to http://0.0.0.0:8080/ifttt/v1/actions/new-file
      1) POST Valid new file from Pryv

  /actions/new-note/
    fields/streamId/options
      ✓ POST Valid token
    /
      ✓ POST  new note with invalid content
      ✓ POST invalid with missing streamId
      ✓ POST invalid with missing description
      ✓ POST invalid with missing tags
      ✓ POST Valid new note (141ms)
      ✓ POST Valid new note 2 (90ms)
      ✓ POST  new note with invalid stream id (91ms)
      ✓ POST  new note valid long content (90ms)
      ✓ POST  new note empty string
      ✓ POST strip tag if too long (191ms)

  /actions/new-numerical-*
    basic
      2) new-numerical-basic/fields/eventType/options
      3) POST Valid new numerical value
      4) POST Invalid new numerical value, missing eventType
      5) POST Invalid new numerical value, missing numericalValue
      6) POST Invalid new numerical value, unparsable numericalValue

  /actions/new-photo/
    /
      7) POST Valid new picture from Pryv

  oauth2
    GET /authorize
      ✓ should redirect to access.html page
      ✓ should have the provided redirectUrl
    POST /token
      Valid token exchange
        ✓ should return a successful response
        ✓ should store the IFTTT and Pryv tokens
        ✓ should create a webhook with the IFTTT token in URL query params (101ms)
      Invalid Token exchange
        ✓ should return an error

  status
    ✓ valid auth
    ✓ invalid auth should fail

  userinfo
    ✓ POST /ifttt/v1/test/setup - Channel Test Setup

  /triggers/new-file/
    fields/streamId/options
      ✓ POST Valid token
    /
      ✓ POST Valid token (166ms)

  /triggers/new-note/
    fields/streamId/options
      ✓ POST Valid token
    /
      ✓ POST Valid token (180ms)
      ✓ POST Valid token urlEndpoint user (135ms)
      ✓ POST Valid token ANY STREAMS (130ms)

  /triggers/new-numerical/
    fields/streamId/options
      ✓ POST Valid token
    fields/eventType/options
      ✓ POST Valid token
    /
      ✓ POST Missing dataType
      ✓ POST Valid token and content (134ms)

  /triggers/new-photo/
    fields/streamId/options
      ✓ POST Valid token
    /
      ✓ POST Valid token (123ms)

  userinfo
    /ifttt/v1/user/info
      ✓ GET /ifttt/v1/user/info - No Authorization header
      old user
        ✓ GET /iftttt/v1/user/info
      new user
        ✓ GET /iftttt/v1/user/info

  webhooks
    POST /webhooks
      Valid call
        ✓ should respond successfully
        ✓ should send the correct headers
        ✓ should have the correct body
      missing IFTTT token
        ✓ should return an error
      invalid IFTTT Token
        ✓ should return an error
      unknown IFTTT token
        ✓ should return an error


  41 passing (2s)
  7 failing

  1) /actions/new-file/ / POST Valid new file from Pryv:

      Uncaught AssertionError: expected 401 to be 200
      + expected - actual

      -401
      +200
      
      at Assertion.fail (node_modules/should/cjs/should.js:275:17)
      at Assertion.value (node_modules/should/cjs/should.js:356:19)
      at dist/test/acceptance/actions/new-file.test.js:34:27
      at Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at parser (node_modules/superagent/lib/node/index.js:916:18)
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)
      at endReadableNT (_stream_readable.js:1064:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  2) /actions/new-numerical-* basic new-numerical-basic/fields/eventType/options:

      Uncaught AssertionError: expected 401 to be 200
      + expected - actual

      -401
      +200
      
      at Assertion.fail (node_modules/should/cjs/should.js:275:17)
      at Assertion.value (node_modules/should/cjs/should.js:356:19)
      at dist/test/acceptance/actions/new-numerical.test.js:90:25
      at Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at parser (node_modules/superagent/lib/node/index.js:916:18)
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)
      at endReadableNT (_stream_readable.js:1064:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  3) /actions/new-numerical-* basic POST Valid new numerical value:

      Uncaught AssertionError: expected 401 to be 200
      + expected - actual

      -401
      +200
      
      at Assertion.fail (node_modules/should/cjs/should.js:275:17)
      at Assertion.value (node_modules/should/cjs/should.js:356:19)
      at dist/test/acceptance/actions/new-numerical.test.js:34:27
      at Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at parser (node_modules/superagent/lib/node/index.js:916:18)
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)
      at endReadableNT (_stream_readable.js:1064:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  4) /actions/new-numerical-* basic POST Invalid new numerical value, missing eventType:

      Uncaught AssertionError: expected 401 to be 400
      + expected - actual

      -401
      +400
      
      at Assertion.fail (node_modules/should/cjs/should.js:275:17)
      at Assertion.value (node_modules/should/cjs/should.js:356:19)
      at dist/test/acceptance/actions/new-numerical.test.js:50:27
      at Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at parser (node_modules/superagent/lib/node/index.js:916:18)
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)
      at endReadableNT (_stream_readable.js:1064:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  5) /actions/new-numerical-* basic POST Invalid new numerical value, missing numericalValue:

      Uncaught AssertionError: expected 401 to be 400
      + expected - actual

      -401
      +400
      
      at Assertion.fail (node_modules/should/cjs/should.js:275:17)
      at Assertion.value (node_modules/should/cjs/should.js:356:19)
      at dist/test/acceptance/actions/new-numerical.test.js:64:27
      at Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at parser (node_modules/superagent/lib/node/index.js:916:18)
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)
      at endReadableNT (_stream_readable.js:1064:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  6) /actions/new-numerical-* basic POST Invalid new numerical value, unparsable numericalValue:

      Uncaught AssertionError: expected 401 to be 400
      + expected - actual

      -401
      +400
      
      at Assertion.fail (node_modules/should/cjs/should.js:275:17)
      at Assertion.value (node_modules/should/cjs/should.js:356:19)
      at dist/test/acceptance/actions/new-numerical.test.js:79:27
      at Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at parser (node_modules/superagent/lib/node/index.js:916:18)
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)
      at endReadableNT (_stream_readable.js:1064:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  7) /actions/new-photo/ / POST Valid new picture from Pryv:
     Error: Unauthorized
      at Request.callback (node_modules/superagent/lib/node/index.js:706:15)
      at parser (node_modules/superagent/lib/node/index.js:916:18)
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)
      at endReadableNT (_stream_readable.js:1064:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)



error Command failed with exit code 7.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
loic@loic-mac 11:54:31 /home/loic/programmation/nodejs/bridge-ifttt $ 
```

File: code/src/middleware/bearerAuth.js (hunk 3 above)
Comment (kebetsi) [reply code/src/middleware/bearerAuth.js]: Using `!` to test for a `null` or `undefined` is commonly used, but depends on JS type coercion which translates them to `false`. I prefer using `== null` which is more readable, although longer to write.

File: code/src/routes/ifttt-v1/actions/start-stop-activity.js (hunk 4 above)
Comment (kebetsi) [reply code/src/routes/ifttt-v1/actions/start-stop-activity.js]: start-stop is not included in the bridge yet. So I left it here.

File: code/src/routes/ifttt-v1/triggers/new-file.js (hunk 5 above)
Comment (kebetsi) [reply code/src/routes/ifttt-v1/triggers/new-file.js]: It is checked just above it.

File: code/src/errors/factory.js (hunk 1 above)
Comment (kebetsi) [reply code/src/errors/factory.js]: yep

File: code/src/fields/stream.js (hunk 2 above)
Comment (kebetsi) [reply code/src/fields/stream.js]: right

File: code/src/routes/ifttt-v1/triggers/new-file.js (hunk 5 above)
Comment (loic-pryv) [reply code/src/routes/ifttt-v1/triggers/new-file.js]: Of course. I should have open the whole file.....
My bad.

File: code/src/middleware/bearerAuth.js (hunk 3 above)
Comment (kebetsi) [reply code/src/middleware/bearerAuth.js]: In addition to readability, it handles correctly cases where you might expect a number `0` or a boolean `false`. For example: 

```
0 == null // false
false == null //false
```

**Pull request merged**